### ISSUE TITLE
console: workspace already unlocked

### ISSUE THREAD

**Issue by carlpett** [hashicorp/terraform#24246]
### Terraform Version
```
Terraform v0.12.21
```

### Terraform Configuration Files
Any file with a syntax error, or don't run `terraform init` to initialize providers.

### Expected Behavior
Running `terraform console` when there are syntax errors or Terraform is uninitialized should error out early, and exit without unrelated error messages.

### Actual Behavior
Terraform seems to always try to release the state lock, even when the lock was never aquired:

```
$ terraform console

Error: Could not satisfy plugin requirements


Plugin reinitialization required. Please run "terraform init".

Plugins are external binaries that Terraform uses to access and manipulate
resources. The configuration provided requires plugins which can't be located,
don't satisfy the version constraints, or are otherwise incompatible.

Terraform automatically discovers provider requirements from your
configuration, including providers used in child modules. To see the
requirements and constraints from each module, run "terraform providers".



Error: provider.helm: no suitable version installed
  version requirements: "~> 1.0,~> 1.0"
  versions installed: none



Error releasing the state lock!

Error message: workspace already unlocked
Lock Info:
  ID:        4abfde4c-063b-8f3a-456c-d84fdccad92b
  Path:
  Operation: OperationTypeInvalid
  Who:       <redacted>
  Version:   0.12.21
  Created:   2020-03-02 09:49:37.542035591 +0000 UTC
  Info:


Terraform acquires a lock when accessing your state to prevent others
running Terraform to potentially modify the state at the same time. An
error occurred while releasing this lock. This could mean that the lock
did or did not release properly. If the lock didn't release properly,
Terraform may not be able to run future commands since it'll appear as if
the lock is held.

In this scenario, please call the "force-unlock" command to unlock the
state manually. This is a very dangerous operation since if it is done
erroneously it could result in two people modifying state at the same time.
Only call this command if you're certain that the unlock above failed and
that no one else is holding a lock.
```

### Steps to Reproduce
0. Don't run `terraform init`
1. `terraform console`

### Additional Context
Using Terraform Cloud for state storage and remote operations.

**Comment by apparentlymart** [comment]
This appears to be an issue specific to the `remote` backend, though its _cause_ seems to be an incorrect assumption in the `terraform console` command implementation.

The `terraform console` command acquires a "context" to do its evaluation in, and the implementation of getting a context is different depending on which backend is selected. It then uses a `defer` to ensure that it always calls `Unlock`  on the operation's `StateLocker`, even if there's an error:

https://github.com/hashicorp/terraform/blob/dc8fd14c1ee9c0f582a8df0a59b27631eff929b0/command/console.go#L93-L102

The problem here, it seems, is that in the case of the remote backend the `Context` method might _not_ exit with an active lock: specifically, it has its _own_ logic to unlock before returning if it encountered any errors:

https://github.com/hashicorp/terraform/blob/dc8fd14c1ee9c0f582a8df0a59b27631eff929b0/backend/remote/backend_context.go#L49-L58

This means that for the remote backend in particular, any errors when creating the context (which includes checking all of the providers to obtain provider schemas) the `op.StateLocker` will be _unlocked_ when the `Context` method returns, which then causes the `terraform console` code above to call `Unlock` when a lock isn't held.

The corresponding case in the _local_ backend (which is, internally, the implementation for every backend _except_ `remote`) returns with an active lock even if it encounters an error, because there is a `Lock` but no corresponding `Unlock` (deferred or otherwise):

https://github.com/hashicorp/terraform/blob/dc8fd14c1ee9c0f582a8df0a59b27631eff929b0/backend/local/backend_local.go#L47-L51

---

To address this, we need to decide what the contract is for `Backend.Context`: if it returns errors, is it the caller's responsibility to unlock? My instinct is to say that it's the `Backend.Context` method's responsibility to guarantee that the lock will be held when returning _only_ if the method is successful (no error diagnostics), because the `Backend.Context` implementation is the only place that can reliably determine whether or not it is holding a lock and thus know whether an unlock is needed.

If we go this route, we'd need to add a corresponding `defer` to the local backend to unlock if there are errors, and then change the `terraform console` implementation to install _its_ `defer` only after it has verified that the `Context` method was successful. That is, to swap the two clauses here:

https://github.com/hashicorp/terraform/blob/dc8fd14c1ee9c0f582a8df0a59b27631eff929b0/command/console.go#L96-L108

We must also take care here that only `ctxDiags.HasErrors()` (not `diags.HasErrors()`) is the decider for whether we must release the lock, because the `Context` method has no awareness of whether there are already diagnostics from earlier operations before it was called.

**Comment by mildwonkey** [comment]
There was an issue with that commit, so it's been reverted. I hope to get a better fix in this week, but wanted to re-open this issue now in case I wasn't able to get it in before the feature freeze.
